vispy/geometry/generation.py

```diff
@@ -89,6 +89,250 @@ def create_cube():
+
+    vertices['position'] = positions
+    vertices['texcoord'] = texcoords
+    vertices['normal'] = normals
+    vertices['color'] = colours
+
+    return vertices, faces, outline
+
+
+def create_box(width=1,
+               height=1,
+               depth=1,
+               width_segments=1,
+               height_segments=1,
+               depth_segments=1,
+               planes=None):
+    """ Generate vertices & indices for a filled and outlined box.
+
+    Parameters
+    ----------
+    width : numeric
+        Box width.
+    height : numeric
+        Box height.
+    depth : numeric
+        Box depth.
+    width_segments : int
+        Box segments count along the width.
+    height_segments : numeric
+        Box segments count along the height.
+    depth_segments : numeric
+        Box segments count along the depth.
+    planes: array_like
+        Any combination of ``{'-x', '+x', '-y', '+y', '-z', '+z'}``
+        Included planes in the box construction.
+
+    Returns
+    -------
+    vertices : array
+        Array of vertices suitable for use as a VertexBuffer.
+    faces : array
+        Indices to use to produce a filled box.
+    outline : array
+        Indices to use to produce an outline of the box.
+    """
+
+    planes = (('+x', '-x', '+y', '-y', '+z', '-z')
+              if planes is None else
+              [d.lower() for d in planes])
+
+    w_s, h_s, d_s = width_segments, height_segments, depth_segments
+
+    planes_m = []
+    if '-z' in planes:
+        planes_m.append(create_plane(width, depth, w_s, d_s, '-z'))
+        planes_m[-1][0]['position'][..., 2] -= height / 2
+    if '+z' in planes:
+        planes_m.append(create_plane(width, depth, w_s, d_s, '+z'))
+        planes_m[-1][0]['position'][..., 2] += height / 2
+
+    if '-y' in planes:
+        planes_m.append(create_plane(height, width, h_s, w_s, '-y'))
+        planes_m[-1][0]['position'][..., 1] -= depth / 2
+    if '+y' in planes:
+        planes_m.append(create_plane(height, width, h_s, w_s, '+y'))
+        planes_m[-1][0]['position'][..., 1] += depth / 2
+
+    if '-x' in planes:
+        planes_m.append(create_plane(depth, height, d_s, h_s, '-x'))
+        planes_m[-1][0]['position'][..., 0] -= width / 2
+    if '+x' in planes:
+        planes_m.append(create_plane(depth, height, d_s, h_s, '+x'))
+        planes_m[-1][0]['position'][..., 0] += width / 2
+
+    positions = np.zeros((0, 3))
+    texcoords = np.zeros((0, 2))
+    normals = np.zeros((0, 3))
+
+    faces = np.zeros((0, 3), dtype=np.uint32)
+    outline = np.zeros((0, 2), dtype=np.uint32)
+
+    offset = 0
+    for vertices_p, faces_p, outline_p in planes_m:
+        positions = np.vstack((positions, vertices_p['position']))
+        texcoords = np.vstack((texcoords, vertices_p['texcoord']))
+        normals = np.vstack((normals, vertices_p['normal']))
+
+        faces = np.vstack((faces, faces_p + offset))
+        outline = np.vstack((outline, outline_p + offset))
```

Comment:
using `vstack` in this way will cause a bunch of unnecessary memory reallocations, it would be better to just make a list of everything and call `np.concatenate` once -- or better, preallocate the proper array size and populate (but this is probably overkill)